Replace Boost.Python with pybind11

CalaW opened the pull request:
This PR replaces Boost.Python with [pybind11](https://github.com/pybind/pybind11), which has become the de facto standard for Python bindings in C++.

Key improvements:

- Simplified binding code, thanks to pybind11’s automatic handling of standard containers.
- Easier build process, as pybind11 is a lightweight, header-only library.
- Additionally, this PR includes fixes for existing CI errors. I'm happy to split those into a separate PR if preferred.

Note: This is still a draft. Documentation updates are pending. Please review the current changes and share your feedback. Once everyone is on board, I’ll proceed with updating the docs.

This should resolve #164.

CalaW committed:
Use pybind11 to replace boost.python

CalaW committed:
Simplify ocl_dropcutter.

Note: Moved BatchDropCutter::getTrianglesUnderCutter to pure cpp implementation

CalaW committed:
fix pip .so installation

CalaW committed:
Cleanup ocl_geometry python binding

CalaW committed:
Fix macos test, default to arm64 if not specified in install.sh args.

CalaW committed:
cleanup install script, remove boost-python

CalaW committed:
Cleanup ocl_algo python binding

CalaW committed:
fix dropcutter get cl points

CalaW committed:
Point inherit in python

CalaW committed:
add stub file, reorder python export to avoid c++ type in docstrings https://pybind11.readthedocs.io/en/latest/advanced/misc.html#avoiding-cpp-types-in-docstrings

CalaW committed:
Revert "Add Boost.Python patch for Python 3.11 support"

This reverts commit e894a138c47fc9e40f11fee18a79188827c6beb4.

CalaW commented:
Hi @aewallin @vespakoen, just following up on this PR. If either of you happens to have some spare time, would you mind giving it a look or testing it out? I’d be happy to make any necessary changes. Thanks in advance!

SpectralVectors commented:
Hi @CalaW - not sure what you needed from this lib, but I built a set of wheels from your branch for Python 3.13 and uploaded them here if you want them: https://github.com/SpectralVectors/opencamlib/releases/tag/Python313_Wheels
Thanks!

CalaW committed:
fix scikit-build

vespakoen commented:
Cool! I will take a look at this next week, I hope to also upgrade Boost to the modern version which should remove a lot of the install script.
I think we still use boost for other stuff, right?
Anyways, i'll take a look soon, thanks for all this effort!

CalaW commented:
Hi @vespakoen, Thanks for taking a look!

Yes, Boost is still required. This PR only replaces Boost.Python with pybind11. We still use Boost.Graph, for example boost/graph/adjacency_list.hpp.

I have already replaced some uses of BOOST_FOREACH, boost::tie, and Boost.Math with the standard library. I can also make another PR to replace the remaining ones.

vespakoen commented:
Removing boost completely would be great, it saves a lot of hassle with their somewhat weird build system.

Boost did gain some better compatibility with CMake though, which might help.
And Boost.Python specifically, was a pain to cross compile, so this will already help a lot!

Maybe getting rid of Boost completely should be a follow up PR, I'd focus on getting this to work on all the platforms.

Will look into this soon...

vespakoen commented:
Did you also consider nanobind? (https://github.com/wjakob/nanobind) it seems to be very similar in syntax so might be worth a try?

vespakoen committed:
Create venv for Python

vespakoen committed:
Install script and nodejs lib fixes

CalaW commented:
Yes, I did consider nanobind. At that time, it was still relatively new, while pybind11 was already the de facto standard and had a much larger user base (3.4k vs 18k star).

From what I have seen, nanobind does have some advantages, especially smaller binaries and lower binding overhead. However, for ocl, I believe most of the heavy computation happens in C++, so I don’t expect the Python binding overhead to be a bottleneck.

vespakoen commented:
I opened a pull request to your fork, If that lands I think I am fine with merging this.
Do you think it's better to do a major release since we changed the binding generation from Boost.Python to pybind11? I guess the API actually stayed the same?
I think it might be a good signal to package maintainers that something bigger changed, since the packagers will have to swap out Boost.Python for pybind11 as well...

Let me know what you all think ;)

EDIT: Just saw that the README also needs some updating

CalaW commented:
Thanks! I will check your PR and update the README as well.

I am not fully sure that the Python API is completely stable, because the project currently does not have a complete test framework for checking API compatibility. I run all python examples and test file and they looked fine. I tried to keep the exposed classes, functions, and method names the same, but there may still be some behavior differences.

For example, `AdaptiveWaterline.getXFibers`, behaves differently in the old Boost.Python bindings. The old binding drops all empty elements before returning the list:

https://github.com/aewallin/opencamlib/blob/95b036fe28ce6d77c97b98e5fbc337904ae49560/src/pythonlib/adaptivewaterline_py.hpp#L50-L60

However, the underlying C++ implementation inherits the behavior from Waterline, which returns the complete list:

https://github.com/aewallin/opencamlib/blob/95b036fe28ce6d77c97b98e5fbc337904ae49560/src/pythonlib/waterline_py.hpp#L53-L61

In the pybind11 version, I preserved the behavior of the C++ implementation instead but it's still easy to modify.

https://github.com/CalaW/opencamlib/blob/5470a8977e2b0540a863c61a3c7b038edfdcf61a/src/algo/waterline.hpp#L54-L55

Speaking of versioning, I noticed that the current versioning seems to be date-based, with the latest release being `2023.01.11`, rather than semantic versioning. In that case, maybe we can make a new dated release and clearly mention that it contains a major binding and packaging change.

CalaW committed:
install script: better support for Python venv and fix Docker argument handling

CalaW committed:
Add py::keep_alive so we can do `waterline.setCutter(ocl.CylCutter())` without segfault

CalaW committed:
remove usage of to_array so pythonlib can use c++14

CalaW committed:
fix: correct syntax in str() override for MillingCutter_py

CalaW committed:
Remove incomplete Python cutter override support

The previous Boost.Python override support for `MillingCutter` was incomplete and effectively unusable:
- `vertexDrop` was not virtual, so Python overrides were never called by `dropCutter()`.
- Several required virtual functions (e.g. height/width queries) were not exposed, making it impossible to implement a complete custom cutter.
- `offsetCutter()` returns an owning raw pointer, which is incompatible with Python object lifetime and could lead to dangling pointers or crashes.

Even if these issues were addressed, overriding the low-level contact routines from Python would not be a practical or useful extension mechanism.

The API was introduced incorrectly and is not known to have been used, so removing it does not introduce a compatibility concern.

CalaW committed:
fix(install): use the active Python environment

Stop creating and activating a project-local virtual environment.
Install and test with the currently selected Python executable instead.

CalaW committed:
change macos libomp bottle tag to sequoia

CalaW committed:
workflow manual run

CalaW committed:
fix(install): update Homebrew libomp fetching for macOS architecture

CalaW committed:
use sequoia instead of tahoe

CalaW committed:
update readme

CalaW committed:
refine binding

- add GIL release
- remove CCType package level export
- change some python constructor signature to match cpp
